### PR TITLE
Balancer controller implementation

### PR DESCRIPTION
**Pull request by mwielgus**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Implementation of the Balancer controller (+autogen client).

### CODE REVIEW

**Review by pbetkier — APPROVED** [review]
/lgtm

**Review by voelzmo — CHANGES_REQUESTED** [review]
Hey @mwielgus I had a look and have a few general questions regarding the structure. Sorry I didn't get to looking at the details of the implementation yet.

* It seems we merged the `balancer` API under [module path `k8s.io/autoscaling`](https://github.com/kubernetes/autoscaler/blob/master/balancer/go.mod#L1), whereas all other components in this repo define modules under `k8s.io/autoscaler`, so I guess we should either correct this first or make it part of this PR?
* I assume `k8s.io/gke-autoscaling` is an internal package location and should rather be `k8s.io/autoscaler` in all occurrences? This is also present in generated files, so I guess these would need to be re-generated afterwards
* Should we add `k8s.io/code-generator` to the dependencies and check in some `hack/update-codegen.sh` script to make it easy to re-generate everything?
* Did you intentionally decide against checking in `vendor`, maybe because of https://github.com/kubernetes/autoscaler/issues/4878? I'm just asking this, because currently all other components in this repo have their `vendor` checked in.

File: balancer/pkg/controller/conditions.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+*/
+
+package controller
+
+import (
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/gke-autoscaling/balancer/pkg/apis/balancer.x-k8s.io/v1alpha1"
```
[the remark this answers is not in the document]
Comment (mwielgus) [reply balancer/pkg/controller/conditions.go]: Fixed.

File: balancer/pkg/controller/core_test.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+		},
+		{
+			name: "With pods, 1 replica, 30/70, with young pending",
+			pods: []*v1.Pod{
+				newPod("default", "b1", v1.PodPending, time.Now().Add(-time.Second*20), "run=b,service=nginx"),
+			},
+			balancer: newBalancer(1),
+			scales: []*hpav1.Scale{
+				newScale("a", 0),
+				newScale("b", 0),
```
[the remark this answers is not in the document]
Comment (mwielgus) [reply balancer/pkg/controller/core_test.go:254]: Scale represents the current reported state of the world. The numbers may be however a bit outdated, the code doesn't care about that number too much.

File: balancer/pkg/controller/core_test.go
[the remark this answers is not in the document]
Comment (voelzmo) [reply balancer/pkg/controller/core_test.go:1]: Thanks for the explanation! This was a bit non-obvious to me in the Priority policy: that _one_ failing Pod indicates that the fallback solution is applied to _all outstanding_ replicas while still trying to place all replicas in A. So this means in the worst case, we would end up temporarily with `2*n-1` running replicas with `n` being the amount of replicas configured in the Balancer, correct?

[the remark this answers is not in the document]
Comment (mwielgus) [reply balancer/pkg/controller/core_test.go:1]: Yes, for a brief moment. As soon as there is N running pods on the first deployment, the second one will be scaled down to 0.

**Comment by grosser** [comment]
some docs would be great, on what this does and how to use it

**Comment by gjtempleton** [comment]
@grosser you can see the initial proposal presenting the problem it aims to address here: https://github.com/kubernetes/autoscaler/blob/master/balancer/proposals/balancer.md

**Pull request merged**